src/db.rs

```diff
@@ -0,0 +1,250 @@
+    };
+
+    console_log!("querying for ({keys:?}) and {}", kind.as_u32());
+
+    // Query for the events first, without the tags
+    let placeholders: String = keys.iter().map(|_| "?").collect::<Vec<_>>().join(", ");
+    let query_str = format!(
+        r#"
+        SELECT * FROM event
+        WHERE pubkey IN ({}) AND kind = ? AND deleted = 0
+        "#,
+        placeholders
+    );
+    let mut stmt = db.prepare(&query_str);
+    let mut bindings = Vec::with_capacity(keys.len() + 1); // +1 for the kind afterwards
+    for key in keys.iter() {
+        bindings.push(JsValue::from_str(key));
+    }
+    bindings.push(JsValue::from_f64(kind.as_u32() as f64));
+    stmt = stmt.bind(&bindings)?;
+
+    let result = stmt.all().await.map_err(|e| {
+        console_log!("Failed to fetch nwc events: {}", e);
+        format!("Failed to fetch nwc events: {}", e)
+    })?;
+
+    let mut events: Vec<Event> = result
+        .results::<Value>()?
+        .iter()
+        .map(|row| {
+            let e: EventRow = serde_json::from_value(row.clone()).map_err(|e| {
+                console_log!("failed to parse event: {}", e);
+                worker::Error::from(format!(
+                    "Failed to deserialize event from row ({}): {}",
+                    row, e
+                ))
+            })?;
+            Ok(Event {
+                id: e.id,
+                pubkey: e.pubkey,
+                created_at: e.created_at,
+                kind: e.kind,
+                tags: e.tags.unwrap_or_default(),
+                content: e.content,
+                sig: e.sig,
+            })
+        })
+        .collect::<Result<Vec<Event>>>()?;
+
+    // Now get all the tags for all the events found
+    let event_ids: Vec<EventId> = events.iter().map(|e| e.id).collect();
+    let placeholders: String = event_ids.iter().map(|_| "?").collect::<Vec<_>>().join(", ");
+    let tag_query_str = format!(
+        r#"
+        SELECT event_id, name, value FROM tag
+        WHERE event_id IN ({})
+        "#,
+        placeholders
+    );
+    let mut tag_stmt = db.prepare(&tag_query_str);
+    let bindings: Vec<JsValue> = event_ids
+        .iter()
+        .map(|id| JsValue::from_str(&id.to_string()))
+        .collect();
+    tag_stmt = tag_stmt.bind(&bindings)?;
+
+    let tag_result = tag_stmt
+        .all()
+        .await
+        .map_err(|e| format!("Failed to fetch tags: {}", e))?;
+
+    let tags: Vec<TagRow> = tag_result.results::<TagRow>()?;
+    let mut tags_map: HashMap<EventId, Vec<Tag>> = HashMap::new();
+
+    for tag_row in tags {
+        if let Ok(tag) = Tag::parse(vec![tag_row.name, tag_row.value]) {
+            tags_map
+                .entry(tag_row.event_id)
+                .or_insert_with(Vec::new)
+                .push(tag);
+        }
+    }
+
+    for event in &mut events {
+        if let Some(tags) = tags_map.remove(&event.id) {
+            event.tags.extend(tags);
+        }
+    }
+
```

Comment:
doing the tags like this might cause the signatures to be invalid, the tags need to be in the same order as before, should at least add a debug_assert to verify the signature is good.

Reply:
Probably just need to sort by id to get in the correct order

Reply:
Nostr is retarded if that's the case

Reply:
1. I log and remove events that don't verify
2. NWC messages should be unaffected since they only have one tag
3. tag inserts happen in order and they are now retrieved by an order by id.